contracts/sol6/bridgeReserve/balancer/KyberBalancerReserve.sol

```diff
@@ -0,0 +1,253 @@
+pragma solidity 0.6.6;
+pragma experimental ABIEncoderV2;
+
+import "./IBalancerExchangeProxy.sol";
+import "../../IKyberReserve.sol";
+import "../../IERC20.sol";
+import "../../utils/Withdrawable3.sol";
+import "../../utils/Utils5.sol";
+import "../../utils/zeppelin/SafeERC20.sol";
+
+
+contract KyberBalancerReserve is IKyberReserve, Withdrawable3, Utils5 {
+    using SafeERC20 for IERC20;
+
+    address public kyberNetwork;
+
+    bool public tradeEnabled = true;
+
+    IBalancerExchangeProxy public exchangeProxy;
+    address public immutable weth;
+
+    // maximum number of pools to use per token
+    // 0 if token is not listed
+    mapping(IERC20 => uint256) public numberPools;
+
+    event TradeExecute(
+        address indexed sender,
+        IERC20 indexed srcToken,
+        uint256 srcAmount,
+        IERC20 indexed destToken,
+        uint256 destAmount,
+        address destAddress
+    );
+
+    event TokenListed(IERC20 indexed token, uint256 numberPools);
+    event TokenDelisted(IERC20 indexed token);
+    event TradeEnabled(bool enable);
+    event EtherReceival(address indexed sender, uint256 amount);
+    event KyberNetworkSet(address kyberNetwork);
+    event BalancerExchangeProxySet(IBalancerExchangeProxy newProxy);
+    event NumberPoolsForTokensUpdated(IERC20[] tokens, uint256[] nPools);
+
+    constructor(
+        IBalancerExchangeProxy _exchangeProxy,
+        address _weth,
+        address _kyberNetwork,
+        address _admin
+    ) public Withdrawable3(_admin) {
+        require(_exchangeProxy != IBalancerExchangeProxy(0), "exchangeProxy 0");
+        require(_weth != address(0), "weth 0");
+        require(_kyberNetwork != address(0), "kyberNetwork 0");
+
+        exchangeProxy = _exchangeProxy;
+        weth = _weth;
+        kyberNetwork = _kyberNetwork;
+    }
+
+    receive() external payable {
+        emit EtherReceival(msg.sender, msg.value);
+    }
+
+    /**
+      conversionRate: expected conversion rate should be >= this value.
+     */
+    function trade(
+        IERC20 srcToken,
+        uint256 srcAmount,
+        IERC20 destToken,
+        address payable destAddress,
+        uint256 conversionRate,
+        bool /* validate */
+    ) external override payable returns (bool) {
+        require(tradeEnabled, "trade is disabled");
+        require(msg.sender == kyberNetwork, "only kyberNetwork");
+
+        require(conversionRate > 0, "conversionRate 0");
```

Comment:
weeelll if we trust the network to give valid rate, this won't happen right =)

Reply:
not totally trust, you can see that we always verify destAmount >= expectedDestAmount
but in scope of bridge reserves, we can safely remove unnecessary check